Built-in Database Migration Tool For Robyn

Charlie-BU opened the pull request:
## Description

This PR adds a built-in database migration tool to Robyn, powered by Alembic. It introduces a new `robyn db` CLI command group that provides a full suite of migration capabilities, including `init`, `migrate`, `upgrade`, `downgrade`, `stamp`, and more.

## Summary

This PR does the following:

- Adds a fully integrated migration CLI (`robyn db`) for SQLAlchemy-based projects
- Wraps Alembic’s command interface with a consistent Robyn-style CLI structure
- Auto-generates and configures `alembic.ini` and `env.py` based on user input or intelligent detection (e.g., database URL, model metadata path)
- Supports common Alembic commands: `init`, `revision`, `migrate`, `upgrade`, `downgrade`, `merge`, `edit`, `stamp`, `show`, `history`, etc.
- Includes support for migration templates (e.g., `--template robyn`)
- Handles command-line errors gracefully with user-friendly messages
- Provides extension class `Migrate` that can be optionally used inside Robyn apps for future integration

This aims to make database versioning and schema evolution first-class citizens in the Robyn developer workflow, reducing friction for backend developers.

<!--
Thank you for contributing to Robyn!
-->

## PR Checklist

Please ensure that:

- [x] The PR contains a descriptive title  
- [x] The PR contains a descriptive summary of the changes  
- [x] You build and test your changes before submitting a PR.  
- [x] You have added relevant documentation  
- [x] You have added relevant tests. We prefer integration tests wherever possible

## Pre-Commit Instructions:

- [x] Ensure that you have run the [pre-commit hooks](https://github.com/sparckles/robyn#%EF%B8%8F-to-develop-locally) on your PR.

Charlie-BU committed:
robyn-db-migrate

Charlie-BU committed:
build: robyn-db-migrate

Charlie-BU committed:
fix: add-db_migration-to-test-cli

Charlie-BU commented on robyn/__init__.py:
```diff
@@ -26,6 +26,14 @@
 from robyn.types import Directory
 from robyn.ws import WebSocket
 
+import importlib.util
+
+spec = importlib.util.find_spec("alembic")
+if spec is not None and spec.loader is not None:
+    from robyn.migrate import Migrate
+else:
+    Migrate = None
+
```
I was considering that user would import this `Migrate` to connect migration with his robyn app, like this:
<img width="3420" height="1960" alt="image" src="https://github.com/user-attachments/assets/99e59c47-a63a-42bd-ae5c-d3cfe2d1a464" />

But later i found a better way, which is using `robyn db ...` directly in cli. So there's no need to export `Migrate` as an extern class for users to call in their `app.py`. So i'll delete it. Thx for finding the problem!

Charlie-BU committed:
fix: remove redundant Migrate class and add special configure for sqlite

sansyrox commented on robyn/cli.py:
```diff
@@ -3,6 +3,7 @@
 import subprocess
 import sys
 import webbrowser
+import argparse
```
do we need this?

Charlie-BU replied:
I think we need this to define a sub-argument for robyn, which is `robyn db ...`. We might need `argparse` to process the argument passed after `robyn db`.
<img width="3420" height="1960" alt="image" src="https://github.com/user-attachments/assets/34b07910-85cf-41e1-afaa-97d05d9c31c1" />

sansyrox commented on robyn/cli.py:
```diff
@@ -71,6 +83,14 @@ def create_robyn_app():
+    if db_migration == "Y":
+        print("Installing the latest version of alembic...")
+        try:
+            subprocess.run([sys.executable, "-m", "pip", "install", "alembic", "-q"], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
```
there should be a better way here, no?

Charlie-BU replied:
I was considering to add `alembic` as a dependency when people execute `pip install robyn`. But in view that not everyone needs this database migration tool, I decided to make it an option for people when they do `robyn create`. If they said yes then `alembic` is a must in this case.
I modified this self-installing process like this:
<img width="3420" height="1960" alt="image" src="https://github.com/user-attachments/assets/860cfa8e-8dc0-4fd0-823f-583e5e222d27" />
Pls check if there's something that could be optimized. Thx! :D

sansyrox replied:
@Charlie-BU , maybe we can allow it to be installed like `pip install robyn[db]` and then it installs alembic too?

Charlie-BU replied:
Cool! I removed this as you said. And set the `pyproject.toml` like this. So people will install alembic when they do `pip install robyn[db]`
<img width="3420" height="1960" alt="image" src="https://github.com/user-attachments/assets/d82b3531-9e8d-4c66-9067-89cac0db3617" />

<img width="3420" height="1960" alt="image" src="https://github.com/user-attachments/assets/5177cd71-0d18-4e8e-add0-ecb1b94ae1de" />

sansyrox commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+        Returns:
+            str: Path to the template directory
+        """
+        if template is None:
+            return Path(__file__).parent / "templates" / "robyn"
+        return Path(template)
```
This path validation might be incorrect.

Charlie-BU replied:
Could you possibly explain more? I think it works.

sansyrox commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+
+
+def _auto_configure_migrations(directory: str, db_url: Optional[str] = None, model_path: Optional[str] = None) -> None:
+    """自动配置 alembic.ini 和 env.py 文件
+
+    Args:
+        directory: 迁移文件目录
+        db_url: 数据库 URL
+        model_path: 模型文件路径
+    """
```
please use english here.

sansyrox commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+            with open(env_py_path, "r") as f:
+                content = f.read()
+            try:
+                content = content.replace(
```
same with this replace.

Charlie-BU replied:
solved :D. Using regex instead.

sansyrox commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+                else:
+                    print('Cannot find models module.\nPlease provide your database URL with "--db-url=<YOUR_DB_URL>".')
+                    return
+        except Exception:
```
let's not use a general `Exception`.

sansyrox commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+            else:
+                print('Cannot find models module.\nPlease provide your model path with "--model-path=<YOUR_MODEL_PATH>".')
+                return
+        except Exception:
```
general catch all

Charlie-BU replied:
solved :D

sansyrox commented:
Hey @Charlie-BU 👋 

Thanks for the PR. This is a good start :D But I have some inline comments.

Charlie-BU commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+
+                # Replace the import statement and target_metadata setting
+                # Allow importing from the parent directory
+                import_statement = f"sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))\nfrom {module_path} import {class_name}\ntarget_metadata = {class_name}.metadata"
+
+                # Replace the import statement
+                content = content.replace(
+                    "# add your model's MetaData object here\n# for 'autogenerate' support\n# from myapp import mymodel\n# target_metadata = mymodel.Base.metadata",
+                    f"# add your model's MetaData object here\n# for 'autogenerate' support\n{import_statement}",
+                )
+
+                # Replace the target_metadata setting
+                content = content.replace(
+                    "target_metadata = config.attributes.get('sqlalchemy.metadata', None)",
+                    "# target_metadata = config.attributes.get('sqlalchemy.metadata', None)\n# Already set by the import above",
```
This `env.py` file and `alembic.ini` file are automatically created by alembic when handling `alembic init alembic`, which is `robyn db init` now. And we have to modify the configuration here to adapt it to the real case.

sansyrox replied:
@Charlie-BU  , so do we need to commit this file?

Charlie-BU replied:
yeah we do. That's a template whose path is needed passing into `alembic init alembic`, which is `command.init(config, directory, template=template_path, package=package)` in our case. 
<img width="3420" height="1960" alt="image" src="https://github.com/user-attachments/assets/e14c2d52-7eec-4bc6-83ec-9cc3275ee53d" />

So that's why I wrote a `_get_template_path()` method in class `Config`. It's a must when we initialize the alembic using python code instead of command `alembic init alembic`.
<img width="3420" height="1960" alt="image" src="https://github.com/user-attachments/assets/6c4c4a81-d684-462f-bfef-705c42e58b7d" />

Charlie-BU commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+                content = f.read()
+
+            # Replace the database URL
+            content = content.replace("sqlalchemy.url = driver://user:pass@localhost/dbname", f"sqlalchemy.url = {db_url}")
```
solved :D. Using regex instead.

Charlie-BU commented on robyn/migrate.py:
```diff
@@ -0,0 +1,624 @@
+        return Path(template)
+
+
+def catch_errors(f):
+    """Decorator to catch and handle errors."""
+
+    @wraps(f)
+    def wrapped(*args, **kwargs):
+        try:
+            return f(*args, **kwargs)
+        except Exception as e:
+            print(f"ERROR: {str(e)}")
+            sys.exit(1)
+
+    return wrapped
```
Let me think how to refactor this...😊

Charlie-BU replied:
I optimized my decorator. Let it catch specific exception while executing alembic commands with `command.*`, and inform the exception detail. Thanks for your advice!!:D

Charlie-BU committed:
fix: db-migration

Charlie-BU committed:
fix: catch-import-error

sansyrox commented on robyn/cli.py:
```diff
@@ -71,6 +83,36 @@ def create_robyn_app():
         os.remove(f"{final_project_dir_path}/Dockerfile")
 
+    # If database migration is needed, install alembic
+    if db_migration == "Y":
+        print("Installing alembic...")
+        try:
+            # Check if alembic is already installed
+            import importlib.util
+
+            alembic_spec = importlib.util.find_spec("alembic")
+
+            if alembic_spec is None:
+                # Install alembic using pip API
+                try:
+                    import pip
+
+                    print("Installing alembic using pip API...")
+                    from pip._internal.cli.main import main as pip_main
+
+                    pip_main(["install", "alembic", "--quiet"])
+                    print("Successfully installed alembic.")
+                except ImportError:
+                    # If pip API is not available, use subprocess
+                    print("Installing alembic using subprocess...")
+                    subprocess.run([sys.executable, "-m", "pip", "install", "alembic", "-q"], check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+                    print("Successfully installed alembic.")
+            else:
+                print("Alembic is already installed.")
+        except (subprocess.CalledProcessError, ImportError) as e:
+            print(f"Failed to install alembic: {str(e)}")
+            print("Please install it manually using 'pip install alembic'.")
+
```
I think we can get rid of this if we decide to use alembic as an optional dependency.

Charlie-BU replied:
NP. Then people gotta install alembic as they do `pip install robyn`. Right?

sansyrox commented on robyn/cli.py:
```diff
@@ -101,6 +143,70 @@ def start_app_normally(config: Config):
     subprocess.run(command, start_new_session=False)
 
 
+def handle_db_command():
```
we are importing a lot of things in this function. let's do that on the top

Charlie-BU replied:
Sure!

sansyrox commented on robyn/templates/robyn/env.py:
is this file autogenerated @Charlie-BU ?

Charlie-BU replied:
yeah, autogenerated by `alembic init alembic`.

Charlie-BU commented on robyn/migrate.py:
I'm really sorrrrrrrry I'm not quite familiar about to design a separate module. I don't know if i need to redesign the whole robyn-db-migration structure to do that?

Charlie-BU committed:
refactor(migration): improve error handling and type safety in migration module

- Replace generic error decorator with typed migration-specific error handling
- Add proper type hints and validation for migration operations
- Rename database dependency key from "database" to "db" for consistency
- Remove redundant db_migration option from CLI
- Improve error messages and logging for migration operations

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sansyrox reviewed (commented):
Thank you so much for this thorough piece of work, @Charlie-BU — there's clearly a lot of care here. My main hesitation is scope: Robyn is intentionally ORM-agnostic and doesn't ship a database layer, so bundling a SQLAlchemy/Alembic-specific migration CLI is a big surface area for us to own long-term, and it largely overlaps with Alembic's own CLI. There's also a blocking correctness issue: `robyn/cli.py` now imports `robyn/migrate.py` unconditionally, and `migrate.py` calls `exit(1)` at import time when Alembic isn't installed — that would break every `robyn` command (create/dev/run) for users who don't have the optional `db` extra. I'd also rather avoid `pip._internal` and runtime auto-installs, and the regex-based rewriting of `env.py`/`alembic.ini` is quite brittle. Given the scope question plus the missing tests, I'm inclined to recommend we close this in favor of documenting an Alembic recipe instead, but I'm very happy to discuss if you feel strongly about it being first-class.

Charlie-BU commented:
@sansyrox  Thanks for reviewing my unnatural idea. I'd like to discuss your suggestion but unfortunately i'm really in a busy time now. If it's okay for you i'd like to open another PR about alembic documentation in a few months.
